Sto email notifications doc 2800 (#1065)
harness/developer-hub: docs/security-testing-orchestration/use-sto/email-notifications.md

---
title: Notification workflows -- Automated emails for detected issues
description: Send emails and stop pipelines automatically based on severity levels
sidebar_position: 210
---

This topic describes two example workflows for setting up automatic notifications based on scan results. 

> When you run a scan, the step generates a set of output variables that capture the number of issues detected at each severity level: CRITICAL, HIGH, MEDIUM, and so on. You can [include these variables in automated notifications](#send-notification-emails-with-scan-results) such as emails. You can also [create governance policies that evaluate these variables](#stop-builds-and-generate-notifications-using-governance-policies) and use the results to stop pipeline builds and send automated notifications.

Reviewer: ```suggestion
When you run a scan, the step generates a set of output variables that capture the number of issues detected at each severity level: CRITICAL, HIGH, MEDIUM, and so on. You can [include these variables in automated notifications](#send-notification-emails-with-scan-results), such as emails. You can also [create governance policies that evaluate these variables](#stop-builds-and-generate-notifications-using-governance-policies) and use the results to stop pipeline builds and send automated notifications.
```

> When you run a scan, the step generates a set of of output variables that capture the number of issues detected at each severity level: CRITICAL, HIGH, MEDIUM, and so on. You can [include these variables in automated notifications](#send-notification-emails-with-scan-results) such as emails. You can also [create governance policies that evaluate these variables](#stop-builds-and-generate-notifications-using-governance-policies) and use the results to stop pipeline builds and send automated notifications.

Reviewer 2: `a set of of output variables` - of repeated twice

## Send notifications with scan results

> In this workflow, you add an Email step that notifies recipients of the scan results. The email lists the number of issues detected, grouped by severity level: critical, high, medium, and so on. The pipeline sends a notification whenever the previous scan step finishes successfully.

Reviewer: In the 2nd paragraph, the severity levels were capitalized but here, they are lowercase.

> In this workflow, you add an Email step that notifies recipients of the the scan results. The email lists the number of issues detected, grouped by severity level: critical, high, medium, and so on. The pipeline sends a notification whenever the previous scan step finishes successfully.

Reviewer 2: `recipients of the the scan results` - the is repeated twice

      type: Email
      name: emailOnDetectedIssues
      identifier: emailOnDetectedIssues
      spec:
        to: your.email@your-org.org
        cc: devs@your-org.org
        subject: "STO ALERT: Scan results for <+pipeline.name>"
        body: |- 
             "STO scan of <+pipeline.name> found the following issues:  <br>
             Critical : <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.CRITICAL> <br>
             New Critical : <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.NEW_CRITICAL> <br>
             High: <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.HIGH> <br>
             New High: <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.NEW_HIGH> <br>
             Medium: <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.MEDIUM> <br>
             New Medium: <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.NEW_MEDIUM> <br> 
             See https://app.harness.io/ng/#/account/ACCOUNT_ID/sto/orgs/default/"
      timeout: 1d
```

> For more information about pipeline and other variables, go to [Built-in and Custom Harness Variables Reference](/docs/platform/Variables-and-Expressions/harness-variables).

Reviewer: ```suggestion
For more information about pipeline variables and other variables, go to [Built-in and Custom Harness Variables Reference](/docs/platform/Variables-and-Expressions/harness-variables).
```

## Stop builds and generate notifications using governance policies

:::note

> This workflow requires a basic knowledge of governance policies and how to implement them using [Harness Policy as Code](/docs/platform/Policy-as-code/harness-governance-overview) and [Open Policy Agent (OPA)](https://www.openpolicyagent.org/)

Reviewer: ```suggestion
This workflow requires a basic knowledge of governance policies and how to implement them using [Harness Policy as Code](/docs/platform/Policy-as-code/harness-governance-overview) and [Open Policy Agent (OPA)](https://www.openpolicyagent.org/).
```

:::

> In this workflow, you create a simple OPA policy for the pipeline: If the scan detected any NEW_CRITICAL or NEW_HIGH severities, exit the build with an error and send an email. "New-severity" issues are issues that were found in the scanned target but not the baseline —  or, if the target has no baseline, the previous scan. 

Reviewer: ```suggestion
In this workflow, you create a simple OPA policy for the pipeline: If the scan detected any NEW_CRITICAL or NEW_HIGH severities, exit the build with an error and send an email. _New-severity_ issues are issues that were found in the scanned target but not the baseline —  or, if the target has no baseline, the previous scan. 
```

### Create the OPA policy

1. Go to **Account Settings** (left menu) &gt; **Policies**.

> 2. Click **Policies** (top right) and then **New Policiy**. 

Reviewer: ```suggestion
2. Select **Policies** (top right) and then **New Policy**. 
```

> 3. Give your new policy the name **Security no NEW_CRITICAL or NEW_HIGH issues**. 

Reviewer: ```suggestion
3. Name the policy **Security no NEW_CRITICAL or NEW_HIGH issues**. 
```

> 3. Given your new policy the name **Security no NEW_CRITICAL or NEW_HIGH issues**. 

Reviewer 2: Should be `Give`

4. In the Edit Policy window, enter the following OPA code: 

   ```
   package pipeline_environment

   deny[sprintf("Scan results can't include any NEW_CRITICAL vulnerabilities '%s'", [input])] {
       input.NEW_CRITICAL != 0 
   }

   deny[sprintf("Scan results can't contain any NEW_HIGH vulnerabilities '%s'", [input])] {
       input.NEW_HIGH != 0 
   }
   ```

5. Click **Save**. 

### Create a policy set

> A _policy set_ is simply a collection of one or more policies. You combine policies into a set and then include it in a Policy step. 

Reviewer: ```suggestion
A _policy set_ is a collection of one or more policies. You combine policies into a set and then include it in a Policy step. 
```

> 1. In the Policies Overview page, click **Policy Sets** (top right) and then **New Policiy Set**. 

Reviewer: ```suggestion
1. In the Policies Overview page, select **Policy Sets** (top right) and then **New Policy Set**. 
```

> 2. Click **New Policy Set**. The Policy Set wizard appears.

Reviewer: ```suggestion
2. Select **New Policy Set**. The Policy Set wizard appears.
```

Is this a duplicate step or do you select **New Policy set** twice?

3. Overview:

    1. Name = **Security Set - Block on Issue Severity**

    2. Entity type this policy applies to = **Custom**

    3. On what event should the policy be set to = **On Step**. 
   
       These settings allow you to apply the member policies to a specific step, which you'll define below. 

4. Policy evaluation criteria: 
 

>    1. Click **Add Policy**.

Reviewer: ```suggestion
   1. Select **Add Policy**.
```

    
   2. Select the policy you just created and set the pull-down to **Error and Exit**. This is the action to take if any policies in the set are violated. 

     ![](./static/notif-opa-select-policy.png)

>    3. Click **Apply** to add the policy to the set, then **Finish** to close the Policy Set wizard.

Reviewer: ```suggestion
   3. Select **Apply** to add the policy to the set, then **Finish** to close the Policy Set wizard.
```

   4. :exclamation: In the **Policy Sets** page, enable  **Enforced** for your new policy set.

      ![](./static/notif-opa-policy-set-enforced-yes.png)

### Add a step to evaluate your policy against the scan results

> Now you will create a Policy step that applies the policy you just defined against the output variables from the scan. If the scan detected any NEW_CRITICAL or NEW_HIGH issues, the pipeline fails. 

Reviewer: ```suggestion
Now you will create a Policy step that applies the policy set you just defined against the output variables from the scan. If the scan detected any NEW_CRITICAL or NEW_HIGH issues, the pipeline fails. 
```

1. Add a **Custom** stage after the stage with the scan step, if you don't have one already. 

2. Add a **Policy** step to the stage.

3. Configure the step as follows:

>    1. Entity Type = **Custom**
> 
>    2. Policy Set — Add the policy set you just created. 
> 
>    3. Payload — Enter the following payload. Make sure that you replace `SCAN_STAGE_ID` and `SCAN_STEP_ID` with the stage and step IDs in your pipeline

Reviewer: ```suggestion
   1. Entity Type = **Custom**

   2. Policy Set = Add the policy set you just created. 

   3. Payload = Enter the following payload. Make sure that you replace `SCAN_STAGE_ID` and `SCAN_STEP_ID` with the stage and step IDs in your pipeline
```

      ```json
      {
      "NEW_CRITICAL": <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.NEW_CRITICAL>, 
      "NEW_HIGH": <+pipeline.stages.SCAN_STAGE_ID.spec.execution.steps.SCAN_STEP_ID.output.outputVariables.NEW_HIGH>
      }
      ```

### Set up email notifications for pipeline failures

You have a Policy that fails the pipeline based on an OPA policy. Now you can configure the stage to send an email notification automatically whenever the pipeline fails. 

> 1. Click **Notifications** (right-side menu). The New Notification wizard appears.

Reviewer: ```suggestion
1. Select **Notifications** (right-side menu). The New Notification wizard appears.
```

2. Set up the notification as follows:

>      1. Overview page —  Enter a notifcation name such as **Pipeline failed -- NEW_CRITICAL or NEW_HIGH issues detected**.

Reviewer: ```suggestion
     1. Overview page —  Enter a notification name, such as **Pipeline failed - NEW_CRITICAL or NEW_HIGH issues detected**.
```

There is a mix of = and - in lists in this topic.

>      2. Pipeline Events page  —  Select **Stage Failed**, for the event that triggers the notification, and the stage that has the Policy step you just created.

Reviewer: Confused by the comma usage. Sentence is not clear.

                    to: john.smithh@myorg.org
                    cc: ""
                    subject: "STO ALERT: Issues found in <+pipeline.name>"
                    body: |-
                      "STO scan of <+pipeline.name> found the following issues: <br> 
                       Critical : <+pipeline.stages.banditScanStage.spec.execution.steps.Bandit_1.output.outputVariables.CRITICAL> <br>
                       New Critical : <+pipeline.stages.banditScanStage.spec.execution.steps.Bandit_1.output.outputVariables.NEW_CRITICAL> <br>
                       High: <+pipeline.stages.banditScanStage.spec.execution.steps.Bandit_1.output.outputVariables.HIGH> &#10; <br>
                       New High: <+pipeline.stages.banditScanStage.spec.execution.steps.Bandit_1.output.outputVariables.NEW_HIGH> <br>
                       Medium: <+pipeline.stages.banditScanStage.spec.execution.steps.Bandit_1.output.outputVariables.MEDIUM> <br>
                       New Medium: <+pipeline.stages.banditScanStage.spec.execution.steps.Bandit_1.output.outputVariables.NEW_MEDIUM>  <br>
                       See https://app.harness.io/ng/#/account/H5W8iol5TNWc4G9h5A2MXg/sto/orgs/default/"

>                   timeout: 1d

Reviewer 2: I'd replace hardcoded account Id with something like <Your Account ID> to make it clear they need to modify it
